### ISSUE TITLE
How to implement on multiple tiered project

### ISSUE THREAD

**Issue by hoquem1** [simpleinjector/SimpleInjector#418]
I am very new to the simple injector. I have used the Ninject on a seperate project where I have a WebApi project, that has a contract class project, with a business layer class project and a data layer class project. Each of those project will need to have their DI defined. With Ninject, I used the NinjectModule and override the Load and was able to define my Bind<Interfacexyx, ConcreteXYZ>. 

How can I do the same thing for each class project with Simple Injector? 

Anyone has any example?

**Comment by dotnetjunkie** [comment]
When practicing Dependency Injection, there is one unique location in the application where object graphs are composed. This location is called the [Composition Root](http://blog.ploeh.dk/2011/07/28/CompositionRoot/) and it is located as close to the application's entry point as possible.

What this means is that _only_ your startup project (in your case the Web API project) should have a dependency on the DI Container. None of the other projects (such as your business layer) should have a dependency on Simple Injector (or any other container).

When putting all registrations in the Composition Root, you can still group registrations into different methods or even classes, but the need for dynamically loaded modules or [packages](https://simpleinjector.readthedocs.io/en/latest/howto.html#package-registrations), as Simple Injector calls them, diminishes. It diminishes, because when all 'modules that are part in the composition are known at compile time (which is typically just a hand-full), we don't need reflection to find them. Or to state the Simple Injector documentation, modules or packaging features exist:

> to accommodate applications that require plug-in like modularization, where parts of the application, packed with their own container registrations, can be independently compiled into a dll and ‘dropped’ into a folder, where the main application can pick them up, without the need for the main application to be recompiled and redeployed.

In other words, unless you need truly drop-in behavior, which you don't seem to need, there's no need to use modules or packages anyway (even if you would locate them in different assemblies).

Long story short, even though Simple Injector contains [Packaging](https://www.nuget.org/packages/SimpleInjector.Packaging/) as logical equivalent of Ninject's Modules feature, you don't need it, because you should place all registrations in your start-up project(s) and when grouping in valuable, you can place them in separate classes, but even then it makes more sense to call them statically instead of by the use of reflection.

**Comment by christophedemey** [comment]
@hoquem1 if your looking for an example you could have a look at this repo i created for learning this amazing framework ;-)
https://github.com/christophedemey/DI-SimpleInjector-Tests

**Comment by hoquem1** [comment]
Hi guys, It is my fault that I didn't explain the architecture very well.  So here is my second attempt. 

Contract Project is the Topmost layer (has IBiz , IDataLayer etc)
Under that I have 3 seperate class projects each References ONLY the Contract  project. There is not dependency between these 3 projects. Each of these projects has to be self containted, so when Business Logic project serving Business class it should know which concreate class to send. Same goes for other 2 projects. And that is why I need to be able to have self contained dependency resolver and not rely on a linear project structure. @christophedemey @dotnetjunkie is this clear at all? I can do in further detail if you want. I was trying to get rid of my Ninject and use Simple Injector in its place. 

Please feel free to ask me questions. I really appreciate your help.

**Comment by dotnetjunkie** [comment]
Hi @hoquem1,

The architectural structure of your application is irrelevant to question of how and where to compose object graphs, i.e. make registrations to your container and resolve from the container. The answer will be the same. Your projects or layers can be self-contained, but that doesn't mean they must (or should) do their own composition. Again, only the startup project should do composition.

Take a look at [this Stackoverflow question](https://stackoverflow.com/questions/9501604/ioc-di-why-do-i-have-to-reference-all-layers-assemblies-in-entry-application) and especially at the answers from both [Mark Seemann](http://blog.ploeh.dk/about/) and myself that describe this concept more in detail.
